src/ai/rival-team-gen.ts

```diff
@@ -0,0 +1,309 @@
+): SpeciesId | undefined {
+  // Drop readonly because typescript does not preserve it with its `isArray` signature
+  const pull1 = pool[roll1] as Mutable<(typeof pool)[number]>;
+  if (!Array.isArray(pull1)) {
+    return pull1;
+  }
+  if (roll2 == null) {
+    return;
+  }
+
+  return pull1[roll2];
+}
+
+/**
+ * Calculates the types that the given species is weak to
+ *
+ * @remarks
+ * - Considers Levitate as a guaranteed ability if the species has it as all 3 possible abilities.
+ * - Accounts for type effectiveness challenges via {@linkcode applyChallenges}
+ *
+ * @privateRemarks
+ * Despite potentially being a useful utility method, this is intentionally *not*
+ * exported because it uses logic specific to this file, such as excluding the second type for Tera starters
+ *
+ * @param species - The species to calculate weaknesses for
+ * @param exclude2ndType - (default `false`) Whether to exclude the second type when calculating weaknesses
+ *  Intended to be used for starters since they will terastallize to their primary type.
+ * @returns The set of types that the species is weak to
+ */
+function getWeakTypes(species: PokemonSpecies, exclude2ndType = false): Set<PokemonType> {
+  const weaknesses = new Set<PokemonType>();
+  const alwaysHasLevitate =
+    species.ability1 === AbilityId.LEVITATE
+    && (species.ability2 == null || species.ability2 !== AbilityId.LEVITATE)
+    && (species.abilityHidden == null || species.abilityHidden !== AbilityId.LEVITATE);
+  for (const ty of getEnumValues(PokemonType)) {
+    if (
+      ty === PokemonType.UNKNOWN
+      || ty === PokemonType.STELLAR
+      || ty > MAX_POKEMON_TYPE
+      || (ty === PokemonType.GROUND && alwaysHasLevitate)
+    ) {
+      continue;
+    }
+    const multiplier = new NumberHolder(getTypeDamageMultiplier(ty, species.type1));
+    applyChallenges(ChallengeType.TYPE_EFFECTIVENESS, multiplier);
+
+    if (multiplier.value >= 2 && !exclude2ndType) {
+      const type2 = species.type2;
+      if (type2 != null) {
+        const multiplier2 = new NumberHolder(getTypeDamageMultiplier(ty, type2));
+        applyChallenges(ChallengeType.TYPE_EFFECTIVENESS, multiplier2);
+        multiplier.value *= multiplier2.value;
+      }
+    }
+    if (multiplier.value >= 2) {
+      weaknesses.add(ty);
+    }
+  }
+
+  return weaknesses;
+}
+
+/**
+ * Calculate the existing types and weaknesses in the party up to the target slot
+ * @param targetSlot - The slot we are calculating up to (exclusive)
+ * @param config - The current rival pool configuration
+ * @param referenceConfig - (default {@linkcode RIVAL_6_POOL}); The reference rival pool configuration to use for type considerations
+ * @param pokemonTypes - A set that will hold the types present in the party
+ * @param pokemonWeaknesses - A map that will hold the weaknesses present in the party, and their counts
+ * @param balanceWeaknesses - (default `false`) Whether include weakness balancing
+ */
+
```

Comment:
The docs should probably specify what "type/weakness balancing" means.
```suggestion
 * Calculate the existing types and weaknesses in the party up to the target slot
 * @remarks
 * At least one of either `balanceTypes` or `balanceWeaknesses` should be `true`,
 * otherwise the function does nothing.
 * @param targetSlot - The slot we are calculating up to (exclusive)
 * @param pokemonTypes - A map that will hold the types present in the party
 * @param pokemonWeaknesses - A map that will hold the weaknesses present in the party, and their counts
 * @param balanceTypes - (Default `false`) Whether to include type balancing
 * @param balanceWeaknesses - (Default `false`) Whether to include weakness balancing
 * @param referenceConfig - (Default {@linkcode RIVAL_6_POOL}); The reference rival pool configuration to use for type considerations
 */
```